### ISSUE TITLE
Allow Overriding the Test Name

### ISSUE THREAD

**Issue by spottedmahn** [microsoft/testfx#410]
## Description
It would be very handy if we could control the `TestName` so we could create nice human-readable test names that show up in the test explorer.

Test frameworks like Mocha support this 😊.  [One example.](https://codeburst.io/how-to-test-javascript-with-mocha-the-basics-80132324752e)

Reference: [SO Post](https://stackoverflow.com/questions/16314845/mstest-how-to-change-name-of-testmethod-in-testexplorer)

## Expected behavior
```csharp
[TestMethod("My user friendly test name with special characters like | or * or \\ etc.")]
public void MyTest()
{
    ...
}
```

or

```csharp
[TestMethod()]
[TestName("My user friendly test name with special characters like | or * or \\ etc.")]
public void MyTest()
{
    ...
}
```

## Actual behavior
The method name is the test name 😪.

**Comment by spottedmahn** [comment]
Is the code that decides what to display in Visual Studio in this repo?  I would imagine not.  I would imagine, it is part the of Visual Studio code base?

**Comment by spottedmahn** [comment]
**Example:** Test name from [From testx repo](https://github.com/Microsoft/testfx/blob/3e4ca97f631eb04587ec62d889998fb08f5b2043/test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Execution/TestMethodRunnerTests.cs#L138)

`ExecuteShouldSkipTestAndSkipFillingIgnoreMessageIfIgnoreAttributeIsPresentOnTestClassButHasNoMessage`  

**vs**

`Execute should skip this test and skip filling IgnoreMessage If IgnoreAttribute is present on the TestClass but has no message`

which one is easier for you to read? 😜

**Comment by cltshivash** [comment]
@spottedmahn : This repo has the code for the test framework and decides the name that will be decided in VS. The ask is valid and we will look into it. we are open for contributions as well.

**Comment by spottedmahn** [comment]
>How do I point Visual Studio at my new custom assemblies?

nvm, I think I found it (now that I'm a full pc and not my phone).

[Contribution Guide](https://github.com/Microsoft/testfx-docs/blob/master/docs/contribute.md)

>This article will help you build, test and consume local builds of the MSTest Framework and Adapter.

**Comment by spottedmahn** [comment]
Making progress!!

![image](https://user-images.githubusercontent.com/14079228/40891919-bebb29cc-675c-11e8-913f-483183b8e34d.png)

**Comment by spottedmahn** [comment]
A few issues (and I'm sure there will be more)

- If I put a '.' in my `DisplayName` the test explorer window **strips** the left hand side
   - i.e. "This is a test w/ a period.  Something else" I get "  Something Else"
- I had to hack it up when the test is passed back from VS.  [Source](https://github.com/spottedmahn/testfx/blob/friendly-test-names/src/Adapter/MSTest.CoreAdapter/Extensions/TestCaseExtensions.cs#L26)

```csharp
internal static UnitTestElement ToUnitTestElement(this TestCase testCase, string source)
{
	var isAsync = (testCase.GetPropertyValue(Constants.AsyncTestProperty) as bool?) ?? false;
	var testClassName = testCase.GetPropertyValue(Constants.TestClassNameProperty) as string;

	// method name from fully qualified name, feels hacky
	var parts = testCase.FullyQualifiedName.Split('.');
	var methodName = parts[parts.Length - 1];
	TestMethod testMethod = new TestMethod(methodName, testClassName, source, isAsync);

	UnitTestElement testElement = new UnitTestElement(testMethod)
	{
		IsAsync = isAsync,
		TestCategory = testCase.GetPropertyValue(Constants.TestCategoryProperty) as string[],
		Priority = testCase.GetPropertyValue(Constants.PriorityProperty) as int?
	};

	return testElement;
}
```

**Comment by jayaranigarg** [comment]
@spottedmahn : 
- We will figure out the '.' part once your PR is out.
- Can you tell us the scenario where earlier logic of testcase.DisplayName is not working for you? TestCase displayname is already constructed in a manner what you are trying to achieve. Refer [this](https://github.com/Microsoft/testfx/blob/master/src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs#L93)

**Comment by spottedmahn** [comment]
>We will figure out the '.' part once your PR is out.

Great 🎉!

---

>Can you tell us the scenario where earlier logic of testcase.DisplayName is not working for you? 

Hmm... not 100% what you are referring to.  I'm going to assume you're referring to:

>I had to hack it up when the test is passed back from VS.

When I run a test from the Test Explorer window, it is unable to find the test:

>Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.TypeInspectionException: 'Method UnitTestProject2.UnitTest1.Say What, You Have Spaces in Your C# Test Names??!! does not exist.'

The problem `ToUnitTestElement()` is using the `DisplayName` as the `TestMethod.Name`:

https://github.com/Microsoft/testfx/blob/e13ca085694aa9a9908afb3e9a90a58841b5fe03/src/Adapter/MSTest.CoreAdapter/Extensions/TestCaseExtensions.cs#L26

**Comment by spottedmahn** [comment]
Hi @jayaranigarg - any updates?

**Comment by jayaranigarg** [comment]
@spottedmahn : Apologies for the delay.

>Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.TypeInspectionException: 'Method UnitTestProject2.UnitTest1.Say What, You Have Spaces in Your C# Test Names??!! does not exist.'

This exception is generated from [here](https://github.com/Microsoft/testfx/blob/master/src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs#L621). This effectively means there is some discrepancy in testMethod.Name

I believe we will have to fix DisplayName in TestCase at the time of creation of testcase object [here]
(https://github.com/Microsoft/testfx/blob/master/src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs#L93).

**Comment by spottedmahn** [comment]
Hi @jayaranigarg - PR submitted... sorry it took so long...

**Comment by spottedmahn** [comment]
implemented via #466

**Comment by ahmedalejo** [comment]
>Just making sure everyone sees this solution for #644

**Found** a **better and simpler extension point** that could be baked into `TestMethodAttribute` with ease.

This is what i´m currently using:

cheers.
```csharp
public class PrettyTestClassAttribute : TestClassAttribute
{
    public override TestMethodAttribute GetTestMethodAttribute(TestMethodAttribute wrappedTestMethodAttribute)
    {
        var attribute = base.GetTestMethodAttribute(wrappedTestMethodAttribute);

        return attribute is PrettyTestMethodAttribute
            ? attribute
            : new PrettyTestMethodAttribute(attribute);
    }
}

public class PrettyTestMethodAttribute : TestMethodAttribute
{
    private readonly TestMethodAttribute wrappedTestMethodAttribute;

    public PrettyTestMethodAttribute(){ }

    public PrettyTestMethodAttribute(TestMethodAttribute wrappedTestMethodAttribute) =>
        this.wrappedTestMethodAttribute = wrappedTestMethodAttribute;

    public override TestResult[] Execute(ITestMethod testMethod)
    {
        TestResult[] results;

        if (wrappedTestMethodAttribute is null)
            results = base.Execute(testMethod);
        else
            results = wrappedTestMethodAttribute.Execute(testMethod);

        if(results.Any())
            results[0].DisplayName = testMethod.TestMethodName
                .Replace("_eq_", " == ")
                .Replace("_neq_", " != ")
                .Replace("_gt_", " > ")
                .Replace("_gte_", " >= ")
                .Replace("_lt_", " < ")
                .Replace("_lte_", " <= ")
                .Replace("Bug_", "🐞");
        return results;
    }
}
```
Hopefully will create a new pull request suggesting to bake this into `TestMethodAttribute`

**Comment by spottedmahn** [comment]
Hi @ahmedalejo - nice!  Please share a link to the new PR once created, thanks!

**Comment by nohwnd** [comment]
@ahmedalejo Fixed the code above to actually compile and space after bug. That approach is genius. 

```cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Linq;

namespace UnitTestProject84
{
    [PrettyTestClass]
    public class UnitTest1
    {
        [TestMethod]
        public void Bug_Test_method_1()
        {
        }
    }

    public class PrettyTestClassAttribute : TestClassAttribute
    {
        public override TestMethodAttribute GetTestMethodAttribute(TestMethodAttribute wrappedTestMethodAttribute)
        {
            var attribute = base.GetTestMethodAttribute(wrappedTestMethodAttribute);

            return attribute as PrettyTestMethodAttribute ?? new PrettyTestMethodAttribute(attribute);
        }
    }

    public class PrettyTestMethodAttribute : TestMethodAttribute
    {
        private readonly TestMethodAttribute wrappedTestMethodAttribute;

        public PrettyTestMethodAttribute() { }

        public PrettyTestMethodAttribute(TestMethodAttribute wrappedTestMethodAttribute) =>
            this.wrappedTestMethodAttribute = wrappedTestMethodAttribute;

        public override TestResult[] Execute(ITestMethod testMethod)
        {
            TestResult[] results = wrappedTestMethodAttribute is null
                ? base.Execute(testMethod)
                : wrappedTestMethodAttribute.Execute(testMethod);

            if (results.Any())
                results[0].DisplayName = testMethod.TestMethodName
                    .Replace("_eq_", " == ")
                    .Replace("_neq_", " != ")
                    .Replace("_gt_", " > ")
                    .Replace("_gte_", " >= ")
                    .Replace("_lt_", " < ")
                    .Replace("_lte_", " <= ")
                    .Replace("Bug_", "🐞 ")
                    .Replace("_", " ");

            return results;
        }
    }
}

```

**Comment by mdpopescu** [comment]
Nice, I just discovered this!

One observation: the parameter description on the constructor here is incorrect:

/// <param name="displayName">Message specifies reason for ignoring.</param>

It was probably copied from Ignore and hasn't been changed correctly.

**Comment by nohwnd** [comment]
@mdpopescu was fixed in #715
